# Trabajo Practio de Laboratorio Nº 1 : Propagación de Incertidumbre


##  Datos del grupo

*   **Curso:** R4052
*   **Nº de Grupo:** 2
*   **Integrantes:**
  1.   Paul Leandro Encinas Claros
  2.   Franco Ezequiel Ferrario
  3.   Matías Sebastián Marchesi
  4.   Bruno Quiroga

  







# Inicializacion

Se deja un bloque de código para la inicialización de las funciones e imports de bibliotecas necesarias.

In [2]:
# Librerias de pyhon a utilizar
import numpy as np
import math
import IPython
from IPython.display import display, Math
import pandas as pd

In [3]:
# Funciones Auxiliares
from typing import Any

def incert_A(mediciones):
  s = np.std(mediciones, ddof=1, )  # ddof=1 -> grados de libertad v = 1/n-1
  return s/math.sqrt(len(mediciones))

def incert_B(err_rel, nd, ctas, mediciones):
  return (err_rel/100 + nd/ctas) * np.mean(mediciones) / math.sqrt(3)

def incert_Total(uA, uB):
  return math.sqrt(uA**2 + uB**2)

def incert_expandida(uC, k):
  return k * uC

# Procedimiento específico de la medición

## Descripción

Para esta práctica se precisan dos resistencias de valores conocidos. En este caso elegimos los valores de $R_A=R_B=10k\Omega$ y una fuente de tensión simple, que en este caso utilizamos una del laboratorio de electrónica de la UTN FRBA.

A su vez, para la medición de todos los valores utilizamos el multímetro **UT890D+**.

El circuito corresponse a un divisor resistivo como es mostrado en la siguiente imagen:

<div style="text-align: center;">
    <img src="recursos/divisor_resistivo.png" width="400" >
</div>

Se divide la práctica en tres casos que se detallarán a continuación. 

### Caso 1: Medición directa
Para la medición directa habrá que conectar la salida del divisor (paralelo a $R_B$) el multímetro antes mencionado.
Allí se obtendrá el resultado de la tensión de salida, y con la datasheet del multímetro se propagará el error del mismo.

### Caso 2: Medición indirecta
Para la medición indirecta se utilizarán los valores de resistencias y tensión de fuente medidos con el multímetro, luego se los propagará teniendo en cuenta que en la ecuación del divisor resistivo todos tienen influencia sobre el otro, por lo que habrá que hacer las derivadas parciales para obtener la incertidumbre total.

### Caso 3: Especificaciones
Para la medición a partir de las especificaciones de los componentes se utilizarán los valores de resistencias según indica el código de colores del fabricante, y para la fuente, como no se posee la especificación se utilizará el valor medido de la fuente directamente con el multímetro y se lo propagará igual que se hizo en los otros casos.
Nuevamente se harán las derivadas parciales para obtener la incertidumbre total.

### Objetivo

El objetivo de la práctica es comparar los tres casos para comparar las formas de medición y obtener resultados similares a la salida del divisor resistivo.

## Datos

Se deja constancia de los datos utilizados por la cátedra, el caso de k utilizado para la incertidumbre expandida y los datos de error tomados de la datasheet del multímetro **UT890D+**. A su vez, se utilizan los mismos para los futuros códigos en python utilizados en esta práctica

In [4]:
k = 1.65

# Datos del multímetro

## Cuentas del display
ctas = 6000

## DC
err_dc = 0.7
n_dc = 3

## Resistencia
err_r = 0.8
n_r = 3

## Fuente
err_fuente = 1
n_fuente = 2

# Resolución


## Caso 1: Medición directa

Para este método se obtendrá la tensión de salida del circuito utilizando el resultado que muestra el display del multímetro. Luego se hallará su incertidumbre. 

El valor indicado en el display del multímetro fue de $5.062 V$.

Utilizando la datasheet del multímetro **UT890D+** que es el utilizado en la práctica se obtiene la especificación del error para mediciones de continua es:
$$\pm (0.7\% + 3 \ dígitos)$$

Para este caso no tenemos múltiples mediciones así que no tendremos incertidumbre del tipo A.

$$\Delta V= (\frac{\epsilon_r \%}{100} + \frac{n}{c_i}) . V$$

$$\mu_B (V) = \frac{\Delta V}{\sqrt{3}}$$

$$U(V) = K_{95\%} . \mu_B(V)$$

In [5]:
ci_1 = 5062
V_directa = 5.062
uC_V_1 = incert_B(err_dc, n_dc, ci_1, [V_directa])

U_V_1 = incert_expandida(uC_V_1, k)
print("U_V_1:",float(U_V_1))

U_V_1: 0.03661330240595623


Como se ve en la simulación, se obtiene la incertidumbre de la medición:
$$U(V) = 0.03661330240595623 V $$

Y el valor medido en el display como se dijo antes es:
$$V = 5.062 V$$

Por lo que, el resultado quedará expresado de la siguiente manera:

$$V_o = (V \pm U(V))$$

Aplicando la **norma IRAM35050 redondeamos** la incertidumbre en base a los **dos dígitos más significativos** y se aplica el mismo criterio con el valor medido.

<div style="text-align: center;">

<div style="
    display: inline-block;
    border: 2px solid;
    border-radius: 10px;
    padding: 12px 20px;
    margin: 10px 0;
">

$$V_o = (5.062 \pm 0.037) V$$

</div>

</div>

## Caso 2: Medición indirecta

Para este caso se medirá la tensión de salida partiendo de los valores de componentes y de la fuente medidos con el mismo multímetro.

Los datos obtenidos con el multímetro son:
$$R_A = 9.84 k\Omega$$
$$R_B = 9.86 k\Omega$$

$$V_F = 10.07 V$$

Para el caso de las resistencias se usó la escala de $60K\Omega$ y para la tensión la escala de $60V$.

Se obtendrá la incertidumbre de tipo B de cada medición, en este caso usando la expresión con la resolución para más simpleza:

$$\Delta X= \frac{\epsilon_r \%}{100} . X  + n.Res(X)$$

Siendo la resolución el cociente entre la escala y la cantidad de cuentas máximas de la escala. 

$$Res(X) = \frac{X}{c_X}$$

Luego, se obtiene la incertidumbre dividiendo en este caso por raiz de tres por ser una distribución rectangular.

$$\mu_B (X) = \frac{\Delta X}{\sqrt{3}}$$

Lo que se hará ahora es utilizando la ecuación del divisor resistivo obtener la tensión resultante. La ecuación es:

$$V_o = V_F . \frac{R_B}{R_A + R_B}$$

También hay que propagar todos los errores derivando para cada caso.



In [6]:
escR = 60e3
cR = 6000
escV = 60
cV = 6000

#Valores medidos
RA = 9.84e3      # ohm
RB = 9.86e3      # ohm
VF = 10.07       # V

ResR = escR / cR 
ResV = escV / cV 

ci_RA = RA / ResR
ci_RB = RB / ResR
ci_VF = VF / ResV

print(ci_RA, ci_RB, ci_VF)

uB_RA_2 = incert_B(err_r, n_r, ci_RA, [RA])
uB_RB_2 = incert_B(err_r, n_r, ci_RB, [RB])
uB_VF_2 = incert_B(err_dc, n_dc, ci_VF, [VF])

print("uB(RA): ",uB_RA_2)
print("uB(RB): ",uB_RB_2)
print("uB(VF): ",uB_VF_2)

984.0 986.0 1007.0
uB(RA):  62.76952126629612
uB(RB):  62.861897309366455
uB(VF):  0.058017928550865495


Habiendo obtenido las incertidumbres de los tres datos ahora debemos propagar la ecuación, para eso sigo la siguiente fórmula:

$$\Large \mu_c(V_o) = \sqrt{\sum_{i=1}^{N} \Biggl(
\frac
{\partial{V_o}}
{\partial{x_i}}  \, \mu(x_i) 
\Biggr)^2}$$

Aplicándolo sobre:
$$V_o = V_F . \frac{R_B}{R_A + R_B}$$

Se deben realizar las derivadas parciales de:

$$\frac{\partial{V_o}}{\partial{V_F}} = \frac{R_B}{R_A+R_B}$$
$$\frac{\partial{V_o}}{\partial{R_A}} = -\frac{V_F . R_B}{(R_A+R_B)^2}$$
$$\frac{\partial{V_o}}{\partial{R_B}} = \frac{V_F . R_A}{(R_A+R_B)^2}$$

Reemplazando en la fórmula antes mostrada:
$$\mu_c(V_o) = \sqrt{ 
\Biggl(\frac{\partial{V_o}}{\partial{V_F}}  . \mu(V_F) \Biggr)^2 + 
\Biggl(\frac{\partial{V_o}}{\partial{R_A}}  . \mu(R_A) \Biggr)^2 +
\Biggl(\frac{\partial{V_o}}{\partial{R_B}}  . \mu(R_B) \Biggr)^2 }
$$

$$\mu_c(V_o) = \sqrt{ 
\Biggl(\frac{R_B}{R_A+R_B}  . \mu(V_F) \Biggr)^2 + 
\Biggl(-\frac{V_F . R_B}{(R_A+R_B)^2}  . \mu(R_A) \Biggr)^2 +
\Biggl(\frac{V_F . R_A}{(R_A+R_B)^2}  . \mu(R_B) \Biggr)^2 }
$$

Se hace el procedimiento en python para obtener los resultados numéricos:

In [7]:
# Derivadas parciales

dVo_dVF = RB/(RA+RB)
dVo_dRA = -(VF*RB)/(RA+RB)**2
dVo_dRB = (VF*RA)/(RA+RB)**2

# Incertidumbre Combinada
uC_Vo = np.sqrt(
    (dVo_dVF*uB_VF_2)**2 +
    (dVo_dRA*uB_RA_2)**2 +
    (dVo_dRB*uB_RB_2)**2
)

# Incertidumbre Expandida
U_Vo = k * uC_Vo
print("U(Vo) = ",U_Vo)

U(Vo) =  0.06082066499967485


Como se ve en la simulación, se obtiene la incertidumbre de la medición:
$$U(Vo) = 0.06082066499967485 V $$

Y el resultado de hacer:
$$V_o = V_F . \frac{R_B}{R_A + R_B}$$
Con los valores medidos al principio del ejercicio $V_o$ vale:
$$V_o = 5.040111675 V$$

Por lo que, expresándolo según la norma se obtiene:

<div style="text-align: center;">

<div style="
    display: inline-block;
    border: 2px solid;
    border-radius: 10px;
    padding: 12px 20px;
    margin: 10px 0;
">

$$V_o = (5.040 \pm 0.061) V$$

</div>

</div>


## Caso 3: Especificaciones

Para este estudio se medirá la tensión utilizando las especificaciones técnicas de los componentes.

En el caso de las resistencias los valores de las mismas se pueden hayar mediante el código de colores. Para este laboratorio utilizamos resistencias de $10K\Omega$, la tolerancia se obtuvo mediante la cuarta banda, que en este caso es dorada por lo que es al $5\%$.

Esto significa que, el $\Delta R$ se obtiene sin el segundo término de la ecuación ya que no hay un display que muestre los resultados, quedando:
$$\Delta R = (\frac{\epsilon_r \%}{100} + 0) . R$$

Obteniendo la incertidumbre final, teniendo una distribución rectangular:
$$\mu_B(R) = \frac{\Delta R}{\sqrt{3}}$$

Para el caso de la fuente, el modelo de la misma es HY3002D. Según la datasheet el error presentado en el display es de:
$$\pm (1\% + 2 \ dígitos)$$

En nuestro caso, en el display aparecía:
$$V_F = 10 V$$

Resolviendo:
$$\Delta V_F = (\frac{1 \%}{100} + \frac{2}{1000}) . V_F$$

$$\mu_B (V_F) = \frac{\Delta V_F}{\sqrt{3}}$$

$$U(V_F) = K_{95\%} . \mu_B(V_F)$$

Se realizan los cálculos en python:

In [14]:
escV = 60
cV = 6000

RA_esp = 10e3
RB_esp = 10e3
VF = 10.00

ResV = escV / cV 
ci_VF = VF / ResV

uB_RA_3 = incert_B(5, 0, 1, [RA_esp])
uB_RB_3 = incert_B(5, 0, 1, [RB_esp])
uB_VF_3 = incert_B(err_fuente, n_fuente, ci_VF, [VF])

print("uB(RA): ",uB_RA_3, "Ω")
print("uB(RB): ",uB_RB_3, "Ω")
print("uB(VF): ",uB_VF_3, "V")

uB(RA):  288.6751345948129 Ω
uB(RB):  288.6751345948129 Ω
uB(VF):  0.06928203230275509 V


Con los resultados de las incertidumbres obtenidas, ahora hay que propagar. Para eso se reutiliza la propagación realizada para el Caso 2, ya que es la misma fórmula:
$$V_o = V_F . \frac{R_B}{R_A+R_B}$$

Luego de realizar las derivadas parciales, la propagación queda:
$$\mu_c(V_o) = \sqrt{ 
\Biggl(\frac{R_B}{R_A+R_B}  . \mu(V_F) \Biggr)^2 + 
\Biggl(-\frac{V_F . R_B}{(R_A+R_B)^2}  . \mu(R_A) \Biggr)^2 +
\Biggl(\frac{V_F . R_A}{(R_A+R_B)^2}  . \mu(R_B) \Biggr)^2 }
$$

In [15]:
# Derivadas parciales

dVo_dVF = RB_esp/(RA_esp+RB_esp)
dVo_dRA = -(VF*RB_esp)/(RA_esp+RB_esp)**2
dVo_dRB = (VF*RA_esp)/(RA_esp+RB_esp)**2

# Incertidumbre Combinada
uC_Vo_3 = np.sqrt(
    (dVo_dVF*uB_VF_3)**2 +
    (dVo_dRA*uB_RA_3)**2 +
    (dVo_dRB*uB_RB_3)**2
)

# Incertidumbre Expandida
U_Vo_3 = k * uC_Vo_3
print("U(Vo) = ",U_Vo_3)

# Resultado del divisor resistivo
Vo_3 = VF * (RB_esp / (RA_esp+RB_esp))

print("Vo = ", Vo_3)

U(Vo) =  0.1778380583564722
Vo =  5.0


Finalmente, con los resultados obtenidos:
$$U(V_o) = 0.1778380583564722 V$$
$$V_o = 5.0 V$$

Se deja expresada la tensión con su propagación, redondeada según la norma IRAM.



<div style="text-align: center;">

<div style="
    display: inline-block;
    border: 2px solid;
    border-radius: 10px;
    padding: 12px 20px;
    margin: 10px 0;
">

$$V_o = (5.00 \pm 0.18)V$$

</div>

</div>

# Resultados de las mediciones
En esta sección se resumen los resultados obtenidos en una tabla: 

In [16]:
datos = {
    "Caso": [1, 2, 3],
    "Vo": ["5.062", "5.040", "5.00"],
    "U(Vo)": ["0.037", "0.061", "0.18"],
    "Resultado": ["(5.062 ± 0.037)", "(5.040 ± 0.061)", "(5.00 ± 0.18)"],
    "Unidad": ["V", "V", "V"],
}

df = pd.DataFrame(datos)

display(df)

,Caso,Vo,U(Vo),Resultado,Unidad
0,1,5.062,0.037,(5.062 ± 0.037),V
1,2,5.040,0.061,(5.040 ± 0.061),V
2,3,5.00,0.18,(5.00 ± 0.18),V


# Conclusiones

Se analizará cada caso:
1) En este caso, se realiza la medición directamente midiendo la tensión de la salida y propagando el error.
2) Aquí se mide cada componente y la tensión de la fuente y luego se propaga el error con todos esos datos.
3) Por último, aquí se utilizan los datos dados por el fabricante tanto en las resistencias como en la fuente.

El caso **(3)** resulta el menos preciso, dándonos un error de un orden de magnitud más grande que los otros dos casos. O sea que, los valores que nos brinda el fabricante no deben ser utilizados para hacer una medición precisa, sino una estimación.

Luego, el caso **(2)** nos resulta en un error mayor que el caso **(1)** ya que hay que propagar la ecuación del divisor resistivo, y eso implica tener que considerar los errores de las mediciones y sus implicancias entre si, por eso se hace la derivada parcial de cada caso. Eso hace que se acumulen errores del instrumento, consiguiendo el doble de error comparado con el **(1)**.

Por último, queda el caso **(1)**, que posiblemente sea el más preciso, ya que se mide directamente la salida del circuito, y se propaga una vez sola el error por la medición de tensión. No se necesita propagar las resistencias ya que aquí no las estamos midiendo.

Finalmente, vemos para esta práctica que conviene usar el primer método para medir la tensión de salida de un divisor resistivo, siempre y cuando el resultado de la impedancia sea bastante menor que el del multímetro utilizado para medir. En este caso con dos resistencias de $10k\Omega$ no tendremos problema estimando una impedancia de multímetro de $10M\Omega$.